bootloader/src/old_machine_bootloader_linux.c

```diff
@@ -0,0 +1,202 @@
+
+    /* Open the command for reading. */
+    fp = popen(commandLine, "r");
+    if (fp == NULL)
+        error("Failed to run command\n" );
+
+    /* Read the output a line at a time - output it. */
+    char* res = fgets(path, maxOutputLength, fp);
+    if (res == NULL)
+        error("ERROR reading commandline\n");
+
+    /* close */
+    pclose(fp);
+
+    return path;
+}
+
+struct response sendRequest(char* server, char* tunnel, char* data){
+    CURL *curl;
+    CURLcode res;
+    struct response s;
+    curl = curl_easy_init();
+    init_response(&s);
+    if(curl) {
+        curl_easy_setopt(curl, CURLOPT_URL, server);
+        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, writefunc);
+        curl_easy_setopt(curl, CURLOPT_WRITEDATA, &s);
+        if(tunnel != NULL){
+            curl_easy_setopt(curl, CURLOPT_PROXY, tunnel);
+        }
+        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, data);
+        /* Perform the request, res will get the return code */
+        res = curl_easy_perform(curl);
+        /* Check for errors */
+        if(res != CURLE_OK){
+            fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
+            s.ptr = "FAILED";
+            s.len = strlen(s.ptr);
+            return s;
+        }
+
+        /* always cleanup */
+        curl_easy_cleanup(curl);
+    }
+    curl_global_cleanup();
+    return s;
+}
+
+char** getIpAddresses(int *addrCount){
+    char** IPs = NULL;
+    int i = 0;
+
+    struct ifaddrs * ifAddrStruct=NULL;
+    struct ifaddrs * ifa=NULL;
+    void * tmpAddrPtr=NULL;
+    getifaddrs(&ifAddrStruct);
+
+    for (ifa = ifAddrStruct; ifa != NULL; ifa = ifa->ifa_next) {
+        if (!ifa->ifa_addr || i >= maxSize) {
+            continue;
+        }
+        if (ifa->ifa_addr->sa_family == AF_INET) { // check it is IP4
+            // is a valid IP4 Address
+            tmpAddrPtr=&((struct sockaddr_in *)ifa->ifa_addr)->sin_addr;
+            i = i+1;
+            IPs = (char**)realloc(IPs, (i+1)*sizeof(*IPs));
+            IPs[i-1] = (char*)malloc(INET_ADDRSTRLEN);
+            inet_ntop(AF_INET, tmpAddrPtr, IPs[i-1], INET_ADDRSTRLEN);
+        }
+    }
+    if (ifAddrStruct!=NULL) freeifaddrs(ifAddrStruct);
+    *addrCount = i;
+    return IPs;
+}
+
+int ping_island(int argc, char * argv[])
+{
+    // Get system info
+    struct utsname systemInfo;
+
+    // Get glibc version
+    char* glibcVersion = executeCommand("ldd --version");
+    printf("glibc: %s\n", glibcVersion);
+
+    char* osVersion = executeCommand("cat /etc/os-release");
+    // Some old distributions like centos6 on AWS has a different os info path
+    if(!strcmp(osVersion, "")){
+        osVersion = executeCommand("cat /etc/system-release");
+    }
+    printf("Os version: %s \n", osVersion);
+
+    // Get all machine IP's
+    int addrCount = 0;
+    char** IPs = getIpAddresses(&addrCount);
+    char* IPstring = concatenate(addrCount, IPs, "\", \"");
```

Comment:
You don't check for failure return value

Reply:
No need. If it fails and returns `\0` which is equal to empty string

Reply:
No, you return \0, which is 0, which is NULL (but not exactly). So later on, the NULL value is sent off to different places.

Reply:
\0 is cast to char* as NULL, not empty string, but rather null ptr.